llama-index-integrations/indices/llama-index-indices-managed-postgresml/llama_index/indices/managed/postgresml/base.py

```diff
@@ -0,0 +1,198 @@
+from llama_index.core.base.base_retriever import BaseRetriever
+from llama_index.core.data_structs.data_structs import IndexDict, IndexStructType
+from llama_index.core.indices.managed.base import BaseManagedIndex, IndexType
+from llama_index.core.schema import (
+    BaseNode,
+    Document,
+    TextNode,
+)
+from pgml import Collection, Pipeline
+
+_logger = logging.getLogger(__name__)
+
+
+class PostgresMLIndexStruct(IndexDict):
+    """PostgresML Index Struct."""
+
+    @classmethod
+    def get_type(cls) -> IndexStructType:
+        """Get index struct type."""
+        return IndexStructType.POSTGRESML
+
+
+class PostgresMLIndex(BaseManagedIndex):
+    """PostgresML Index.
+
+    The PostgresML index implements a managed index that uses PostgresML as the backend.
+    PostgresML performs a lot of the functions in traditional indexes in the backend:
+    - breaks down a document into chunks (nodes)
+    - Creates the embedding for each chunk (node)
+    - Performs the search for the top k most similar nodes to a query
+    - Optionally can perform text-generation or chat completion
+    """
+
+    def __init__(
+        self,
+        collection_name: str,
+        pipeline_name: Optional[str] = None,
+        pipeline_schema: Optional[Dict[str, Any]] = None,
+        pgml_database_url: Optional[str] = None,
+        show_progress: bool = True,
+        upsert_parallel_batches: int = 1,
+        nodes: Optional[Sequence[BaseNode]] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize the PostgresML SDK."""
+        self.show_progress = show_progress
+        self.upsert_parallel_batches = upsert_parallel_batches
+
+        index_struct = PostgresMLIndexStruct(
+            index_id=collection_name,
+            summary="PostgresML Index",
+        )
+
+        super().__init__(
+            show_progress=show_progress,
+            index_struct=index_struct,
+            **kwargs,
+        )
+
+        # Create our Collection and Pipeline
+        self.collection = Collection(collection_name, pgml_database_url)
+        if pipeline_name is None:
+            pipeline_name = "v1"
+        if pipeline_schema is None:
+            pipeline_schema = {
+                "content": {
+                    "splitter": {
+                        "model": "recursive_character",
+                        "parameters": {"chunk_size": 1500},
+                    },
+                    "semantic_search": {
+                        "model": "intfloat/e5-small-v2",
+                        "parameters": {"prompt": "passage: "},
+                    },
+                }
+            }
+        self.pipeline = Pipeline(pipeline_name, pipeline_schema)
+
+        # We must wrap self.collection.add_pipeline() with this async function
+        # This is a limitation of the pyo3 async implementation
+        async def add_pipeline():
+            await self.collection.add_pipeline(self.pipeline)
+
+        run_async_tasks([add_pipeline()])
```

Comment:
there is a util in `async_utils` called `asyncio_run()`, which if you use, you might be able to remove these little async function defs

Reply:
Unfortunately the async function defs will still be required. Under the hood we are using the pyo3 rust crate to bind to our Rust library, and we can't touch those functions in non async contexts.

Reply:
Really? Interesting. Alright then!